bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java

```diff
@@ -3,41 +3,79 @@
 // Licensed under the Apache License, Version 2.0.
 package org.ethereum.evmc;
 
-import org.ethereum.evmc.EvmcLoaderException;
-
+import java.io.IOException;
 import java.nio.ByteBuffer;
-import java.util.Objects;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
 
 /**
  * The Java interface to the evm instance.
  *
  * <p>Defines the Java methods capable of accessing the evm implementation.
  */
 public final class EvmcVm implements AutoCloseable {
-  private static EvmcVm evmcVm;
-  private static boolean isEvmcLibraryLoaded = false;
+  private static final Throwable errorLoadingEvmc;
   private ByteBuffer nativeVm;
+
+  static {
+    Throwable error = null;
+    try {
+      // load so containing the jni bindings to evmc
+      System.loadLibrary("libevmc-java");
+    } catch (UnsatisfiedLinkError e) {
+      String extension = null;
+      String operSys = System.getProperty("os.name").toLowerCase();
+      if (operSys.contains("win")) {
+        extension = "dll";
+      } else if (operSys.contains("nix") || operSys.contains("nux") || operSys.contains("aix")) {
+        extension = "so";
+      } else if (operSys.contains("mac")) {
+        extension = "dylib";
+      } else {
+        error = e;
+      }
+      if (extension != null) {
+        try {
+          Path evmcLib = Files.createTempFile("libevmc-java", extension);
+          Files.copy(
+              EvmcVm.class.getResourceAsStream("/libevmc-java." + extension),
+              evmcLib,
+              StandardCopyOption.REPLACE_EXISTING);
+          evmcLib.toFile().deleteOnExit();
+          try {
+            System.load(evmcLib.toAbsolutePath().toString());
+          } catch (UnsatisfiedLinkError e1) {
+            error = e1;
```

Comment:
Why is there a need for a local `error` variable? Because `errorLoadingEvmc` can be set once in the function?

Reply:
Yes, it’s final, and therefore immutable.